src/isar/storage/uploader.py

```diff
@@ -22,21 +72,48 @@ def run(self) -> None:
             inspection: Inspection
             mission_metadata: MissionMetadata
             try:
-                inspection, mission_metadata = self.upload_queue.get(timeout=5)
+                if self._internal_upload_queue:
+                    self._process_upload_queue()
+
+                inspection, mission_metadata = self.upload_queue.get(timeout=1)
+
+                # If new item from thread queue, add one per handler to internal queue:
+                for storage_handler in self.storage_handlers:
+                    new_item: UploaderQueueItem = UploaderQueueItem(
+                        inspection, mission_metadata, storage_handler, _retry_count=-1
+                    )
+                    self._internal_upload_queue.append(new_item)
             except Empty:
                 continue
 
-            for storage_handler in self.storage_handlers:
-                try:
-                    storage_handler.store(
-                        inspection=inspection, metadata=mission_metadata
-                    )
-                    self.logger.info(
-                        f"Storage handler: {type(storage_handler).__name__} "
-                        f"uploaded inspection {str(inspection.id)[:8]}"
-                    )
-                except StorageException:
-                    self.logger.warning(
-                        f"Storage handler: {type(storage_handler).__name__} "
-                        f"failed to upload inspection: {str(inspection.id)[:8]}"
-                    )
+    def _upload(self, upload_item: UploaderQueueItem):
+        try:
+            upload_item.storage_handler.store(
+                inspection=upload_item.inspection, metadata=upload_item.mission_metadata
+            )
+            self.logger.info(
+                f"Storage handler: {type(upload_item.storage_handler).__name__} "
+                f"uploaded inspection {str(upload_item.inspection.id)[:8]}"
+            )
+            self._internal_upload_queue.remove(upload_item)
+        except StorageException:
+            if upload_item.get_retry_count() < self.max_retry_attempts:
+                upload_item.increment_retry(self.max_wait_time)
+                self.logger.warning(
+                    f"Storage handler: {type(upload_item.storage_handler).__name__} "
+                    f"failed to upload inspection: {str(upload_item.inspection.id)[:8]}. "
+                    f"Retrying in {upload_item.seconds_until_retry()}s."
+                )
+            else:
+                self.logger.error(
+                    f"Storage handler: {type(upload_item.storage_handler).__name__} "
+                    f"exceeded max retries to upload inspection: {str(upload_item.inspection.id)[:8]}. "
+                    "Aborting upload."
+                )
```

Comment:
If this statement is hit should the upload item be removed from the internal queue?